CRM: 3433 - Additional HPOS tweaks

tbradsha opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Resolves Automattic/zero-bs-crm#3433 - Additional HPOS tweaks
Resolves Automattic/zero-bs-crm#3434 - Fix segment error output

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This does a bit of cleanup around the HPOS changes made in #35797:

* Prevents some PHP fatals caused by our hooks logic.
* Ensures the contact metabox shows on subscription screens.
* Updates the code that gets subscription data for the contact profile tab.
* Uses proper date functions for subscription/membership dates.

Note that the changelog doesn't reflect most of the above, as it's either behind-the-scenes changes or unreleased regressions.

Also, I solved a small segment listview issue here (meant to be on a different branch + PR, but pushed it onto this one, so leaving it).

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
Test on JPCRM + Woo (non-HPOS) as well as JPCRM + Woo (HPOS) to ensure no regressions. HPOS is enabled on all new empty Woo sites; one can disable it per these instructions:

https://woo.com/document/high-performance-order-storage/#section-7

1. Enable WooSync.
2. Disable WooCommerce.
3. Enable WooCommerce.

In `trunk`, this will cause a PHP fatal whether or not an HPOS site.

In the `fix/crm/3433-hpos-tweaks` branch, it'll work as expected.

4. Create a Woo subscription assigned to an existing contact.
5. Ensure the contact metabox shows on the right side of the subscription screen (in `trunk` it won't on an HPOS site).
6. Ensure the subscription shows up on the contact profile. If the WP user exists, it'll show on both `trunk` and this branch. If the WP user doesn't exist, it'll only show on HPOS sites with this branch (as it matches by email).

---
Segment issue:
1. Create a segment using an Advanced Segments condition.
2. Disable the Advanced Segments extension.
3. Visit the Segments listview: `/wp-admin/admin.php?page=manage-segments`

In `trunk`, the error message is improperly escaped:
![image](https://github.com/Automattic/jetpack/assets/32492176/4583994f-3828-4142-9ac8-cde77020dc66)

In the `fix/crm/3433-hpos-tweaks` branch, it is properly escaped:
![image](https://github.com/Automattic/jetpack/assets/32492176/b975e3cc-b672-4c28-8bf4-122b643139a9)

tbradsha committed:
Don't set up Woo Admin hooks if Woo isn't active

tbradsha committed:
Gather screens where we want to add metaboxes

tbradsha committed:
Don't set up local Woo hooks if Woo isn't active

tbradsha committed:
Add backslash to class lookup

tbradsha committed:
Coding standards

tbradsha committed:
Remove legacy condition

tbradsha committed:
Coding standards

tbradsha committed:
Simplify Woo subs code using wc_get_orders()

tbradsha committed:
Further simplify code

tbradsha committed:
Use proper date functions

tbradsha committed:
Add changelog

tbradsha committed:
Fix segment listview escape issue

tbradsha committed:
Update changelog

gogdzl commented on projects/plugins/crm/modules/woo-sync/includes/jpcrm-woo-sync-contact-tabs.php:
```diff
@@ -354,142 +352,52 @@ private function get_contact_memberships( $object_id = -1 ){
+	 *
+	 * @param int $object_id Contact ID.
+	 */
+	private function get_contact_subscriptions( $object_id = -1 ) {
```
That was a great improvement! Wow!

gogdzl commented:
> If the WP user doesn't exist, it'll only show on HPOS sites with this branch (as it matches by email).

All is working but this part. I have an HPOS site, where the subscriptions are not showing, even when I add a WP User.

![image](https://github.com/Automattic/jetpack/assets/37049295/b617c869-3cc1-41c4-b4e0-97a29768261d)

![image](https://github.com/Automattic/jetpack/assets/37049295/074563f4-72f7-487f-9e14-dcac302b3707)

![image](https://github.com/Automattic/jetpack/assets/37049295/5fe858be-579d-448f-ba37-f1e63e0c0d83)

![image](https://github.com/Automattic/jetpack/assets/37049295/25420eef-343a-4f6f-9f2b-9b1b995e9543)

tbradsha committed:
Search all subscription statuses

tbradsha committed:
Ensure each email alias adds to the array

tbradsha commented:
> the subscriptions are not showing, even when I add a WP User.

 Thanks for the catch! As it turns out, `wc_get_orders` searches all default Woo statuses, and subscriptions have different statuses. I've updated this in 6c44193.

Also, I fixed an issue in the loop: 38c3530

gogdzl reviewed (approved):
Great work as usual @tbradsha !
Ship it!
